src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -19,11 +20,15 @@ public partial class ApplicationEngine : ExecutionEngine
         private readonly bool testMode;
         private readonly List<NotifyEventArgs> notifications = new List<NotifyEventArgs>();
         private readonly List<IDisposable> disposables = new List<IDisposable>();
+        private readonly List<StorageKey> updatedKeys = new List<StorageKey>();
+        private long maxConsumedGas = 0;
 
         public TriggerType Trigger { get; }
         public IVerifiable ScriptContainer { get; }
         public StoreView Snapshot { get; }
         public long GasConsumed { get; private set; } = 0;
+        public long GasCredit { get; private set; } = 0;
+        public long MinimumGasRequired { get { return Math.Max(GasConsumed, maxConsumedGas); } }
```

Comment:
I think we could remove `MinimumGasRequired` and `maxConsumedGas`, but keep `GasConsumed` and rename `GasCredit` to `RecyclingRewardGas`.
Add reuse and release discount rewards to `RecyclingRewardGas`.  In this way, we don't have to modify the `Wallets.MakeTransaction`. 
After tx.script executed, we can pay back the `RecyclingRewardGas` to user.

Reply:
@Tommo-L  if we remove maxConsumedGas the user will need the full-balance to send the TX. I think it is not possible to do it without updating the wallet, because the user can send less fee now.

Reply:
> I think it is not possible to do it without updating the wallet, because the user can send less fees now.

If we allow user send less fees, we will face the contract halting problem. as the transaction's sysfee will loss the use for limitation of execution, is it right?

Reply:
This minimum is exactly for that. The minimum required for the transaction to not fault.